runtime: Add trusted runtime option

jcvenegas opened the pull request:
Some CRI compatible runtimes may not support provileged operations.
Specifically hypervisor based runtimes (like kata-containers, cc-runtime
and runv) do not support privileged operations like:

Provide access to the host namespaces Create fully privileged containers
with access to host devices.  Hypervisor based runtimes create container
workloads within virtual machines. When running host privileged
containers with them, they wont provide support to requested the
privileged opertations.

This commits add the option to define two runtimes:

trusted runtime : Used when a privileged container is requested.
Default runtime : for non-privileged workloads.

A container that belongs to a privileged pod will inherent this property
an will be created with the trusted runtime.

- Add options to define trusted runtime
- Add logic to decide if a sanbox is trusted
- Export annotation containers below to a trusted sandbox

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

Random-Liu committed:
Report containerd version instead of cri-containerd version.

Signed-off-by: Lantao Liu <lantaol@google.com>

yanxuean committed:
some comments

Signed-off-by: yanxuean <yan.xuean@zte.com.cn>

jcvenegas commented:
@sameo @egernst  ping please take a look

jcvenegas commented:
New passing cri-tools using mixed runtimes:
```
CRI_CONTAINERD_FLAGS="--containerd-runtime-engine cc-runtime --containerd-trusted-runtime-engine runc " make test-cri
```
- runtime should ignore a seccomp profile that blocks setting hostname when privileged
- mount with 'rprivate' should not support propagation
 - runtime should support HostPID
- should support propagation from host to container and vice versa
- runtime should support HostIpc is true
- mount with 'rslave' should support propagation from host to container	
 - HostNetwork is true

sameo commented on pkg/server/container_create.go:
```diff
@@ -77,6 +77,7 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
 	config := r.GetConfig()
 	sandboxConfig := r.GetSandboxConfig()
 	sandbox, err := c.sandboxStore.Get(r.GetPodSandboxId())
+
```
Nit: Unneeded change.

sameo commented on cmd/cri-containerd/options/options.go:
```diff
@@ -45,13 +45,21 @@ type ContainerdConfig struct {
 	// Runtime is the runtime to use in containerd. We may support
 	// other runtimes in the future.
 	Runtime string `toml:"runtime" json:"runtime,omitempty"`
+	// TrustedRuntime is the runtime to use for trusted workloads in containerd.
```
`with containerd`

sameo commented on cmd/cri-containerd/options/options.go:
```diff
@@ -45,13 +45,21 @@ type ContainerdConfig struct {
 	// other runtimes in the future.
 	Runtime string `toml:"runtime" json:"runtime,omitempty"`
+	// TrustedRuntime is the runtime to use for trusted workloads in containerd.
+	TrustedRuntime string `toml:"runtime" json:"runtime,omitempty"`
```
Here we should specify that `TrustedRuntime` is optional and when not specified will be either left empty or set to the same value as `Runtime`.

sameo commented on cmd/cri-containerd/options/options.go:
```diff
@@ -45,13 +45,21 @@ type ContainerdConfig struct {
 	// We may support other runtime engines in the future.
 	RuntimeEngine string `toml:"runtime_engine" json:"runtimeEngine,omitempty"`
+	// TrustedRuntimeEngine is the name of the runtime engine used by containerd to run trusted workloads.
+	// Containerd default should be "runc"
```
If I remember the convoluted containerd code correctly, `RuntimeEngine` will translate in the containerd container runtime path or simply binary name. So I think here we should add more comments to specify that the cri-containerd default will be `runc` but that this should be used to specify alternative values like `kata-runtime` or `cc-runtime`.
The whole containerd runtime concept is complex enough and we should try to add the minimal amount of complexity with yet another runtime layer.

sameo commented on pkg/annotations/annotations.go:
```diff
@@ -31,4 +31,7 @@ const (
 	// SandboxID is the sandbox ID annotation
 	SandboxID = "io.kubernetes.cri.sandbox-id"
+
+	// PrivilegedSandboxis the privileged annotation
```
typo: s/PrivilegedSandboxis/PrivilegedSandbox is/

sameo reviewed (commented):
A few minor comments, overall looks quite good.

yanxuean committed:
Handle containerd event reliably

fix #434

Signed-off-by: yanxuean <yan.xuean@zte.com.cn>

Random-Liu committed:
Add CLI package for `ctr` to use.

Signed-off-by: Lantao Liu <lantaol@google.com>

Commit:
Add document about registry configuration

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

stevvooe committed:
pkg/store: use a sync.Once to synchronize channel close

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Random-Liu committed:
Merge pull request #558 from Random-Liu/report-containerd-version

Report containerd version instead of cri-containerd version.

Random-Liu committed:
Merge pull request #654 from stevvooe/simplify-stopch

pkg/store: use a sync.Once to synchronize channel close

jcvenegas committed:
runtime: Add trusted runtime option

Some CRI compatible runtimes may not support provileged operations.
Specifically hypervisor based runtimes (like kata-containers, cc-runtime
and runv) do not support privileged operations like:

- Provide access to the host namespaces
- Create fully privileged containers with access to host devices

Hypervisor based runtimes create container workloads within virtual machines.
When a running host privileged containers using them,
they wont provide support to requested the privileged opertations.

This commits add the new options to define two runtimes:

Trusted runtime : Used when a privileged container is requested.
Default runtime : for non-privileged workloads.

A container that belongs to a privileged pod will inherent this property
an will be created with the trusted runtime.

- Add options to define trusted runtime
- Add logic to decide if a sanbox is trusted
- Export annotation containers below to a trusted sandbox

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
test: Allow change containerd config

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

Random-Liu committed:
Update containerd to 3013762fc58941e33ba70e8f8d9256911f134124

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Use direct function call.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #651 from miaoyq/add-doc-for-registry-config

Add document about registry configuration

Random-Liu committed:
Merge pull request #658 from Random-Liu/direct-function-call

Direct function call

mikebrow committed:
move links for cri-containerd to cri

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

mikebrow committed:
docs update for cri-containerd to cri move

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

Random-Liu committed:
Merge pull request #661 from mikebrow/rename-cri-containerd

renames cri-containerd to cri

Random-Liu committed:
Merge pull request #630 from Random-Liu/add-cli-package

Add CLI package for `ctr` to use.

abhi committed:
Moving to use go-cni library from containerd

This fix aims to use the cni library form containerd.
The library avoid usage of nsenter.

Signed-off-by: abhi <abhi@docker.com>

abhi committed:
Vendoring changes

Signed-off-by: abhi <abhi@docker.com>

abhi committed:
Modifying fake cni plugin

Signed-off-by: abhi <abhi@docker.com>

Random-Liu committed:
Merge pull request #628 from yanxuean/nits

Handle containerd event reliably

Random-Liu committed:
Cleanup event backoff.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Enable promisc mode because Kubernetes network model needs it.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Update containerd to 3c1ef1a714cf5b0104f340f76d539802fc24c75f

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Replace `ctrcri` with `ctr cri`.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #670 from Random-Liu/enable-promic-mode

Enable promisc mode because Kubernetes network model needs it.

Random-Liu committed:
Merge pull request #669 from Random-Liu/cleanup-event-backoff

Cleanup event backoff.

Random-Liu committed:
Merge pull request #671 from Random-Liu/ctrcri-to-ctr-cri

Ctrcri to ctr cri

Random-Liu committed:
Merge pull request #663 from abhi/cni

Moving to use go-cni library from containerd

Random-Liu committed:
Sync vendors with containerd

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Add build-containerd.sh to build containerd from existing repo

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #673 from Random-Liu/test-containerd-repo

Test containerd repo

abhi committed:
Removing DAD config and updating plugins to v0.7.0

Signed-off-by: abhi <abhi@docker.com>

Random-Liu committed:
Fix containerd repo test.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #667 from mikebrow/doc-updates-cri-move

docs update for cri-containerd to cri move

Random-Liu committed:
Merge pull request #674 from abhi/master

Removing DAD config and updating plugins to v0.7.0

Random-Liu committed:
Merge pull request #675 from Random-Liu/fix-containerd-repo-test

Fix containerd repo test.

Random-Liu committed:
Write all instance metadata to files.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Use `github.com/pkg/errors`

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #678 from Random-Liu/use-pkg-errors

Use `github.com/pkg/errors`

Random-Liu committed:
Merge pull request #677 from Random-Liu/all-metadata-to-file

Write all instance metadata to files.

Random-Liu committed:
Only set deploy path.

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #680 from Random-Liu/fix-containerd-test

Only set deploy dir.

Random-Liu committed:
Rename all variables to remove "cricontainerd".

Signed-off-by: Lantao Liu <lantaol@google.com>

Random-Liu committed:
Merge pull request #683 from Random-Liu/rename-variables

Rename all variables to remove "cricontainerd".

jcvenegas committed:
test: Add unit tests for privileged runtime functions

- Add unit test for privilegedSandbox

- Add unit test  for getRuntime

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>
